d2/src/main/java/com/linkedin/d2/discovery/stores/zk/ZooKeeperEphemeralStore.java

```diff
@@ -90,6 +94,8 @@
   private final int _zookeeperReadWindowMs;
   private final ZookeeperChildFilter _zookeeperChildFilter;
   private final ZookeeperEphemeralPrefixGenerator _prefixGenerator;
+  private D2ServiceDiscoveryEventHelper _eventHelper;
+  private volatile ZookeeperNodePathAndDataCallback _znodePathAndDataCallback;
```

Comment:
Use `AtomicReference`?

Reply:
the callback is only set at setStore (line 552 in ZookeeperAnnouncer), which is only called when the store is started: (ZooKeeperConnectionManager.java, line 329). Would that be a problem?

Reply:
I share the same concern, yet I saw that volatile has been used in multiple other variables, and reading some docs online, I realized that Java's volatile is different from C's volatile. In C, it is almost never legit to use volatile while in Java, volatile can be used a lot to achieve eventual consistency.

Reply:
My personal preference is AtomicReference, because it is by default safer without much reasoning :)

Reply:
yeah, makes sense. Changed this and other usage of volatile to AtomicReference.